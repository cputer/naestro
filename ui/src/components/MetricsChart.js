import React, { useEffect, useState } from 'react';
import { LineChart, Line, CartesianGrid, XAxis, YAxis, Tooltip, Legend, ResponsiveContainer } from 'recharts';
import io from 'socket.io-client';

const MetricsChart = () => {
  const [data, setData] = useState([]);

  useEffect(() => {
    const socket = io('http://localhost:4000');
    socket.on('metrics', (metrics) => {
      setData((prev) => [...prev.slice(-19), metrics]);
    });
    return () => socket.close();
  }, []);

  return (
    <ResponsiveContainer width="100%" height={300}>
      <LineChart data={data}>
        <CartesianGrid stroke="#ccc" />
        <XAxis dataKey="time" />
        <YAxis />
        <Tooltip />
        <Legend />
        <Line type="monotone" dataKey="latency" stroke="#8884d8" />
        <Line type="monotone" dataKey="throughput" stroke="#82ca9d" />
      </LineChart>
    </ResponsiveContainer>
  );
};

export default MetricsChart;
