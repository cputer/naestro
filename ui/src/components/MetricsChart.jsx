import React, { useEffect, useState } from "react";
import {
  LineChart,
  Line,
  CartesianGrid,
  XAxis,
  YAxis,
  Tooltip,
  Legend,
  ResponsiveContainer,
} from "recharts";
import io from "socket.io-client";
import { SOCKET_BASE_URL } from "../config";

/**
 * MetricsChart renders latency/throughput metrics.
 *
 * It optionally accepts a `getMetrics` function which can deterministically
 * push metrics to the chart (used in tests or for alternative data sources).
 * When omitted the component will subscribe to the default socket stream.
 *
 * @param {{getMetrics?: (emit: (m: any) => void) => (void|(() => void))}} props
 */
const MetricsChart = ({ getMetrics }) => {
  const [data, setData] = useState([]);
  const [error, setError] = useState(false);

  useEffect(() => {
    let cleanup = () => {};

    const emit = (metrics) => {
      if (Array.isArray(metrics)) {
        metrics.forEach((m) =>
          setData((prev) => [...prev.slice(-19), m])
        );
      } else if (metrics) {
        setData((prev) => [...prev.slice(-19), metrics]);
      }
    };

    const setup = async () => {
      try {
        if (getMetrics) {
          const result = await getMetrics(emit);
          if (typeof result === "function") cleanup = result;
        } else {
          const socket = io(SOCKET_BASE_URL);
          socket.on("metrics", emit);
          socket.on("connect_error", () => setError(true));
          cleanup = () => socket.close();
        }
      } catch {
        setError(true);
      }
    };

    setup();

    return () => {
      cleanup();
    };
  }, [getMetrics]);

  if (error) {
    return <div role="alert">Error loading metrics</div>;
  }

  return (
    <ResponsiveContainer width="100%" height={300}>
      <LineChart data={data}>
        <CartesianGrid stroke="#ccc" />
        <XAxis dataKey="time" />
        <YAxis />
        <Tooltip />
        <Legend />
        <Line type="monotone" dataKey="latency" stroke="#8884d8" />
        <Line type="monotone" dataKey="throughput" stroke="#82ca9d" />
      </LineChart>
    </ResponsiveContainer>
  );
};

export default MetricsChart;
