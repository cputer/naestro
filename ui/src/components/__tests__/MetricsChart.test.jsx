import { render, act, waitFor } from '@testing-library/react';
import { describe, it, expect, vi, beforeEach } from 'vitest';
import MetricsChart from '../MetricsChart.jsx';

const listeners = {};

vi.mock('socket.io-client', () => ({
  default: vi.fn(() => ({
    on: vi.fn((event, cb) => {
      listeners[event] = cb;
    }),
    close: vi.fn(),
  })),
}));

vi.mock('recharts', () => ({
  ResponsiveContainer: ({ children }) => (
    <div data-testid="container">{children}</div>
  ),
  LineChart: ({ data, children }) => (
    <div role="chart" data-points={data.length}>
      {children}
    </div>
  ),
  CartesianGrid: () => <div role="grid" />,
  XAxis: () => <div role="x-axis" />,
  YAxis: () => <div role="y-axis" />,
  Tooltip: () => <div role="tooltip" />,
  Legend: () => <div role="legend" />,
  Line: ({ dataKey }) => <div role={`line-${dataKey}`} />,
}));

describe('MetricsChart', () => {
  beforeEach(() => {
    for (const key in listeners) delete listeners[key];
  });

  it('renders empty response', async () => {
    const getMetrics = vi.fn((emit) => emit([]));
    const { getByRole } = render(<MetricsChart getMetrics={getMetrics} />);

    await waitFor(() => {
      expect(getByRole('chart').getAttribute('data-points')).toBe('0');
    });
  });

  it('renders metrics with axes and legend', async () => {
    const metrics = [{ time: 1, latency: 2, throughput: 3 }];
    const getMetrics = vi.fn((emit) => emit(metrics));
    const { getByRole, container } = render(
      <MetricsChart getMetrics={getMetrics} />
    );

    await waitFor(() => {
      expect(getByRole('chart').getAttribute('data-points')).toBe('1');
    });

    expect(getByRole('x-axis')).toBeInTheDocument();
    expect(getByRole('legend')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('handles socket errors', () => {
    const { getByRole } = render(<MetricsChart />);
    act(() => {
      listeners.connect_error(new Error('fail'));
    });
    expect(getByRole('alert')).toBeInTheDocument();
  });
});

