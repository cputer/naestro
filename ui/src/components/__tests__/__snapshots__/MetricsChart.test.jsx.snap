// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MetricsChart > renders metrics with axes and legend 1`] = `
<div>
  <div
    data-testid="container"
  >
    <div
      data-points="1"
      role="chart"
    >
      <div
        role="grid"
      />
      <div
        role="x-axis"
      />
      <div
        role="y-axis"
      />
      <div
        role="tooltip"
      />
      <div
        role="legend"
      />
      <div
        role="line-latency"
      />
      <div
        role="line-throughput"
      />
    </div>
  </div>
</div>
`;
